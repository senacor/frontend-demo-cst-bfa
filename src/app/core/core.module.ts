import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule } from '@angular/common/http';
import { CcVerificationService } from './services/cc-verification.service';

@NgModule({
  declarations: [],
  imports: [
    CommonModule,
    HttpClientModule,
  ],
  exports: [],
  providers: [
    CcVerificationService,
  ],
})
export class CoreModule {
  constructor(@Optional() @SkipSelf() core:CoreModule) {
    if (core) {
      throw new Error('You should import core module only in the root module!');
    }
  }
}
