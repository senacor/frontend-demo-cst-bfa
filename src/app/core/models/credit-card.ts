export type CreditCard = {
  holder: string;
  ccNumber: string;
  expiration: string;
  ccv: string;
};
