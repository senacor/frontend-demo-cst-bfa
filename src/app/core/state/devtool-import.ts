/* eslint-disable import/no-extraneous-dependencies */
import { StoreDevtoolsModule } from '@ngrx/store-devtools';
import { environment } from '../../../environment/environment';

export const StoreDevModules = [
  StoreDevtoolsModule.instrument({ maxAge: 25, logOnly: environment.production }),
];
