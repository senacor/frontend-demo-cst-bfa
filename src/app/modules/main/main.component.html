<div class="relative flex min-h-screen flex-col justify-center overflow-hidden bg-gray-50 py-6 sm:py-12">
  <div class="absolute inset-0 bg-center [mask-image:linear-gradient(180deg,white,rgba(255,255,255,0))]"></div>
  <div class="relative bg-white px-6 pt-10 pb-8 shadow-xl ring-1 ring-gray-900/5 sm:mx-auto sm:max-w-lg sm:rounded-lg sm:px-10">
    <div class="mx-auto max-w-md">
      <div class="divide-y divide-gray-300/50">
        Frontend @ BA Example
        <div class="space-y-6 py-8 text-base leading-7 text-gray-600">
          <p>This is a basic demonstration of the technologies that are being employed at a governmental agency. In this example we provide a credit card form, pass data between the form and the main page using ngrx for state management, use cypress for E2E tests and so forth...</p>

          <div data-cy="credit-card-information">
            <div *ngIf="(creditCard | async) === undefined">
              No credit card information available
            </div>
            <div *ngIf="(creditCard | async) !== undefined">
              <p class="space-y-1">Credit Card Information</p>
              <br/>
              <ul class="space-y-4">
                <li class="flex items-center">
                  <svg class="h-6 w-6 flex-none fill-sky-100 stroke-sky-500 stroke-2" stroke-linecap="round" stroke-linejoin="round">
                    <circle cx="12" cy="12" r="11" />
                    <path d="m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9" fill="none" />
                  </svg>
                  <p class="ml-4" data-cy="cc-account-holder">
                    Account holder: {{ccHolder | async}}
                  </p>
                </li>

                <li class="flex items-center">
                  <svg class="h-6 w-6 flex-none fill-sky-100 stroke-sky-500 stroke-2" stroke-linecap="round" stroke-linejoin="round">
                    <circle cx="12" cy="12" r="11" />
                    <path d="m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9" fill="none" />
                  </svg>
                  <p class="ml-4" data-cy="cc-number">
                    CC #: {{ccNumber | async | creditCardMask}}
                  </p>
                </li>

                <li class="flex items-center">
                  <svg class="h-6 w-6 flex-none fill-sky-100 stroke-sky-500 stroke-2" stroke-linecap="round" stroke-linejoin="round">
                    <circle cx="12" cy="12" r="11" />
                    <path d="m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9" fill="none" />
                  </svg>
                  <p class="ml-4" data-cy="cc-expiration">
                    Expiration: {{ccExpiration | async}}
                  </p>
                </li>

                <li class="flex items-center">
                  <svg class="h-6 w-6 flex-none fill-sky-100 stroke-sky-500 stroke-2" stroke-linecap="round" stroke-linejoin="round">
                    <circle cx="12" cy="12" r="11" />
                    <path d="m8 13 2.165 2.165a1 1 0 0 0 1.521-.126L16 9" fill="none" />
                  </svg>
                  <p class="ml-4" data-cy="cc-ccv">
                    CCV: {{ccVerification | async}}
                  </p>
                </li>

              </ul>

            </div>
          </div>
        </div>
        <div class="pt-8 text-base font-semibold leading-7">
          <button
            type="button"
            class="inline-block rounded text-sky-500 hover:text-sky-600"
            (click)="navigateToCcForm()"
            *ngIf="(creditCard | async) === undefined"
          >
            Enter credit card information &rarr;
          </button>

          <button
            class="inline-block rounded text-sky-500 hover:text-sky-600"
            (click)="deleteCCInformation()"
            *ngIf="(creditCard | async) !== undefined"
          >
            Delete credit card information
          </button>

        </div>
      </div>
    </div>
  </div>
</div>
