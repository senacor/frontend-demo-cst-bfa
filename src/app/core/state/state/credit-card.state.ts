import { CreditCard } from '../../models/credit-card';

export type CreditCardState = {
  creditCard: Readonly<CreditCard | undefined>
};
