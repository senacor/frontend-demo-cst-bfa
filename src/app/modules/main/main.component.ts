import { Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { Router } from '@angular/router';
import {
  selectCreditCard,
  selectCreditCardExpiration,
  selectCreditCardHolder,
  selectCreditCardNumber, selectCreditCardVerification,
} from '../../core/state/selectors/credit-card.selectors';
import { deleteCreditCardInformation } from '../../core/state/actions/credit-card.actions';

@Component({
  selector: 'app-main',
  templateUrl: './main.component.html',
  styleUrls: ['./main.component.scss'],
})
export class MainComponent {
  readonly creditCard = this.store.select(selectCreditCard);

  readonly ccHolder = this.store.select(selectCreditCardHolder);

  readonly ccNumber = this.store.select(selectCreditCardNumber);

  readonly ccExpiration = this.store.select(selectCreditCardExpiration);

  readonly ccVerification = this.store.select(selectCreditCardVerification);

  constructor(
    private store: Store,
    private router: Router,
  ) {}

  navigateToCcForm() {
    this.router.navigate(['main/cc-information']);
  }

  deleteCCInformation() {
    this.store.dispatch(deleteCreditCardInformation());
  }
}
