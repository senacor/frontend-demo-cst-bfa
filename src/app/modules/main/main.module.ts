import { NgModule } from '@angular/core';
import { MainComponent } from './main.component';
import { CommonModule } from '@angular/common';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MainRoutingModule } from './main-routing.module';
import { SharedModule } from '../../shared/shared.module';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    FormsModule,
    ReactiveFormsModule,
    MainRoutingModule,
  ],
  exports: [MainComponent],
  declarations: [MainComponent],
})
export class MainModule {}
