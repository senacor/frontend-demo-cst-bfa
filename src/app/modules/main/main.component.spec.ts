import { ComponentFixture, TestBed } from '@angular/core/testing';
import { MainComponent } from './main.component';
import { expect } from '@jest/globals';
import { provideMockStore } from '@ngrx/store/testing';
import { CreditCardState } from '../../core/state/state/credit-card.state';
import {
  selectCreditCard, selectCreditCardExpiration,
  selectCreditCardFeatureState,
  selectCreditCardHolder, selectCreditCardNumber, selectCreditCardVerification,
} from '../../core/state/selectors/credit-card.selectors';

describe('MainComponent', () => {
  let component: MainComponent;
  let fixture: ComponentFixture<MainComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ MainComponent ],
      providers: [
        provideMockStore<CreditCardState>({
          selectors: [
            { selector: selectCreditCardFeatureState, value: undefined },
            { selector: selectCreditCard, value: undefined },
            { selector: selectCreditCardHolder, value: '' },
            { selector: selectCreditCardNumber, value: '' },
            { selector: selectCreditCardExpiration, value: '' },
            { selector: selectCreditCardVerification, value: '' },
          ],
        }),
      ],
    })
      .compileComponents();

    fixture = TestBed.createComponent(MainComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
    expect(fixture.nativeElement).toMatchSnapshot();
  });
});
