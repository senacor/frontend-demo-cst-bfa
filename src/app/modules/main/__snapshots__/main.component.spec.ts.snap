// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainComponent should create 1`] = `
<div
  id="root0"
>
  <div
    class="relative flex min-h-screen flex-col justify-center overflow-hidden bg-gray-50 py-6 sm:py-12"
  >
    <div
      class="absolute inset-0 bg-center [mask-image:linear-gradient(180deg,white,rgba(255,255,255,0))]"
    />
    <div
      class="relative bg-white px-6 pt-10 pb-8 shadow-xl ring-1 ring-gray-900/5 sm:mx-auto sm:max-w-lg sm:rounded-lg sm:px-10"
    >
      <div
        class="mx-auto max-w-md"
      >
        <div
          class="divide-y divide-gray-300/50"
        >
           Frontend @ BA Example 
          <div
            class="space-y-6 py-8 text-base leading-7 text-gray-600"
          >
            <p>
              This is a basic demonstration of the technologies that are being employed at a governmental agency. In this example we provide a credit card form, pass data between the form and the main page using ngrx for state management, use cypress for E2E tests and so forth...
            </p>
            <div
              data-cy="credit-card-information"
            >
              <div>
                 No credit card information available 
              </div>
              
              
            </div>
          </div>
          <div
            class="pt-8 text-base font-semibold leading-7"
          >
            <button
              class="inline-block rounded text-sky-500 hover:text-sky-600"
              type="button"
            >
               Enter credit card information â†’ 
            </button>
            
            
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
